src/StoreClient.h

```diff
@@ -67,17 +68,23 @@ class store_client
 public:
     store_client(StoreEntry *);
     ~store_client();
-    bool memReaderHasLowerOffset(int64_t) const;
+    /// \returns the smallest of the passed value and the memory offset of the client
+    int memHeaderOffsetLowerThan(const int64_t offset) const;
     int getType() const;
     void fail();
     void callback(ssize_t len, bool error = false);
+    void callbackClientSide();
     void doCopy (StoreEntry *e);
     void readHeader(const char *buf, ssize_t len);
     void readBody(const char *buf, ssize_t len);
     void copy(StoreEntry *, StoreIOBuffer, STCB *, void *);
     void dumpStats(MemBuf * output, int clientNumber) const;
+    /// whether doCopy() can be called
+    bool canCopy() const { return canScheduleCallback() && !flags.disk_io_pending; }
+    /// TODO: revise/verify all usages
+    /// whether callback() can be called
+    bool canScheduleCallback() const { return _callback.pending() && !clientSideCaller; }
```

Comment:
callback() is not documented so it is difficult for a reader to understand what "callback() can be called" really means. I hope we know enough in this area to describe this new method in a more direct way. Please try.

Reply:
This method basically means the same as canCopy(), not caring about disk swapin state. In other words, canCopy() starts returning 'false' as soon as store_client::fileRead() is called, whereas canScheduleCallback() starts returning 'false' a bit later, when the client-side callback is scheduled. Basically, canScheduleCallback() is somewhat an internal StoreClient method, despite it is declared as 'public'. It has the only 'public' usage is from global/static functions storeSwapInFileClosed()->storeSwapInStart(), which in turn are invoked only from store_client::startSwapin().

Reply:
Thank you. I am not sure, and may need more help deciphering this, but it sounds like canScheduleCallback() means "the client is expecting an STCB call, and we have not scheduled one yet".

We can also explain, in the store_client class description, that "the Store client uses this class to be informed (via an STCB call) about StoreEntry data availability and writing aborts", tying the two descriptions together.

Are the above descriptions accurate?

Reply:
>  "the client is expecting an STCB call, and we have not scheduled one yet".

I am not against this description. We should keep in mind that this should cover not only 'scheduling' but also 'start executing', i.e., we cannot schedule the next STCB until the previous one has started.

>  "the Store client uses this class to be informed (via an STCB call) ..."

Agreed.

Reply:
> Code: `_callback.pending() && !clientSideCaller`

> Alex: the client is expecting an STCB call, and we have not scheduled one yet

> Eduard: We should keep in mind that this should cover not only 'scheduling' but also 'executing', i.e., we cannot schedule the next STCB until the previous one has finished.

Does the PR method code express that "must be false until the previous STCB has finished" restriction? If yes, which of the two data members that the method looks at will change when STCB finishes? 

Or do you mean that the callers of this method need to keep that restriction in mind but the method itself is not responsible for enforcing it? Perhaps there is a restriction on this method use then? For example, perhaps it should only be called inside some specific call chain??

Background: My code reviews are a lot more efficient and less error-prone when I know what the methods should be doing (regardless of the current code inside those methods). This is one of the primary reasons why I am trying to understand what this method means to you, its author. I am trying to get as clear/accurate description of the method as possible. The one I proposed looks clear, but your reaction to it seems to imply that my description is inaccurate -- it does not cover some important caveats. I am trying to improve it so that all the important stuff is covered, hopefully without writing a long paragraph to describe a seemingly basic method.

Reply:
> Or do you mean that the callers of this method need to keep that restriction in mind but the method itself is not responsible for enforcing it?

Yes, the method itself does not/cannot distinguish whether the previous STCB call is still running (i.e., the client buffer is still in use), because store_client::_callback becomes nil when STCB starts. According to my JIRA [analysis](https://measurementfactory.atlassian.net/browse/SQUID-527?focusedCommentId=42871&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-42871), the situation, when we start another STCB call while the previous one is in progress, is not possible (currently). New callers, if any, should keep this in mind, of course.